### PR TITLE
Collapsible Container - Connected with UTs

### PR DESCRIPTION
**Pull request by MichelML**
video: https://drive.google.com/file/d/1vgA5992nNksyDjkCm3RSrUGgFpRgBw8M/view
image:
screen shot 2018-01-24 at 2 38 29 pm  

live demo (the style is not good for the icons, but this is just in the live demo): https://coveo.github.io/react-vapor/collapsiblecontainer/

### CODE REVIEW

File: src/components/collapsibleContainer/CollapsibleContainerConnected.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+      <Provider store={store}>
+        <CollapsibleContainerConnected {...props}>
+          dummy children
+        </CollapsibleContainerConnected>
+      </Provider>,
+      { attachTo: document.getElementById('App') },
+    );
+  };
+
+  it('should get what to do onMount', () => {
```
Comment (sbelzile-Coveo) [inline src/components/collapsibleContainer/CollapsibleContainerConnected.spec.tsx]: Le nom du test devrait être qqchose comme `should not throw onMount`. Même chose pour les 2 tests d'après.

File: src/utils/ClassNameUtils.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { IClassName } from './ClassNameUtils';
```
Comment (sbelzile-Coveo) [inline src/utils/ClassNameUtils.ts]: Nécéssaire ce self-import?

File: src/components/collapsibleContainer/CollapsibleContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  onUnmount?: () => void;
+  onToggleExpandedState?: (currentExpandedState: boolean) => void;
+}
+
+export interface ICollapsibleContainerProps extends
+  ICollapsibleContainerOwnProps,
+  ICollapsibleContainerStateProps,
+  ICollapsibleContainerDispatchProps { }
+
+export class CollapsibleContainer extends React.Component<ICollapsibleContainerProps, {}> {
```
[the remark this answers is not in the document]
Comment (MichelML) [reply src/components/collapsibleContainer/CollapsibleContainer.tsx]: fixed

File: src/components/collapsibleContainer/CollapsibleContainerConnected.spec.tsx (hunk 1 above)
Comment (MichelML) [reply src/components/collapsibleContainer/CollapsibleContainerConnected.spec.tsx]: fixed

File: src/utils/ClassNameUtils.ts (hunk 2 above)
Comment (MichelML) [reply src/utils/ClassNameUtils.ts]: fixed

File: src/components/collapsibleContainer/CollapsibleContainer.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+export class CollapsibleContainer extends React.Component<ICollapsibleContainerProps> {
+  private toggleCollapsibleBodyOnExpandedStateChange(nextProps: ICollapsibleContainerProps) {
+    const collapsibleBodySelector = `#${this.props.id} .collapsible-body`;
+    const animationOptions = { duration: 350 };
+
+    if (this.props.expanded && !nextProps.expanded) {
+      $(collapsibleBodySelector).slideUp(animationOptions);
+    } else if (!this.props.expanded && nextProps.expanded) {
+      $(collapsibleBodySelector).slideDown(animationOptions);
```
Comment (sroy3) [inline src/components/collapsibleContainer/CollapsibleContainer.tsx]: Let's try to use jquery only when absolutely necessary. You can easily use css transitions here. You can also use React's transitions https://reactjs.org/docs/animation.html.

File: src/components/collapsibleContainer/CollapsibleContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    }
+  }
+
+  private getCollapsibleHeaderClass(): string {
+    return classNames(
+      'collapsible-header btn',
+      { active: this.props.expanded },
```
Comment (sroy3) [inline src/components/collapsibleContainer/CollapsibleContainer.tsx]: I think `active: ...` should be on its own line just like `display-block` below, or the opposite.

File: src/components/collapsibleContainer/CollapsibleContainer.tsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    const collapsibleHeaderIcon = this.props.informationUrl
+      ? <LinkSvg url={this.props.informationUrl} target='_blank' tooltip={tooltipProps} svg={svgProps} />
+      : <Tooltip {...tooltipProps}><Svg {...svgProps} /></Tooltip>;
+
+    return <span className='round-contextual-help'>{collapsibleHeaderIcon}</span>;
+  }
+
+  componentWillReceiveProps(nextProps: ICollapsibleContainerProps) {
+    this.toggleCollapsibleBodyOnExpandedStateChange(nextProps);
```
Comment (sroy3) [inline src/components/collapsibleContainer/CollapsibleContainer.tsx]: You might want to look if the expanded prop has changed here. Else `this.toggleCollapsibleBodyOnExpandedStateChange(nextProps);` will be called with every prop change.

Comment (MichelML) [reply src/components/collapsibleContainer/CollapsibleContainer.tsx]: i do it in the function itself...that would be duplicate logic

File: src/components/collapsibleContainer/CollapsibleContainer.tsx (hunk 4 above)
Comment (MichelML) [reply src/components/collapsibleContainer/CollapsibleContainer.tsx]: I did try with css and nothing worked as what we have in backbone/marionette...i guess you arrived at the same conclusion for the table collapsible rows where you used jquery as well? I would like you to consider leaving it as is here...this is the exact same behavior we have everywhere in the anim

Comment (MichelML) [reply src/components/collapsibleContainer/CollapsibleContainer.tsx]: We could log another story for that too, i don't see it as a priority as of now. I can do that if you want.

File: src/components/collapsibleContainer/CollapsibleContainer.tsx (hunk 5 above)
Comment (MichelML) [reply src/components/collapsibleContainer/CollapsibleContainer.tsx]: fixed

File: src/components/collapsibleContainer/CollapsibleContainer.tsx (hunk 6 above)
Comment (sroy3) [reply src/components/collapsibleContainer/CollapsibleContainer.tsx]: `toggleCollapsibleBodyOnExpandedStateChange` is still called every time though. You could remove it from the function and bring it here instead.

Comment (MichelML) [reply src/components/collapsibleContainer/CollapsibleContainer.tsx]: if i move it here, the same code will run in both cases...might not be here anyway after the jquery thing

File: src/components/collapsibleContainer/CollapsibleContainerReducers.spec.ts [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+      const setExpandedToTrueAction = setExpandedCollapsibleContainer('a different id', true);
+      const newState = collapsibleContainersReducer(oldState, setExpandedToTrueAction);
+
+      expect(newState[0].expanded).toBeUndefined();
+
+      const setExpandedToFalseAction = setExpandedCollapsibleContainer('a different id', false);
+      const secondNewState = collapsibleContainersReducer(oldState, setExpandedToFalseAction);
+
+      expect(secondNewState[0].expanded).toBeUndefined();
+    });
```
Comment (GermainBergeron) [inline src/components/collapsibleContainer/CollapsibleContainerReducers.spec.ts]: You should probably also validate that your actions aren't modifying the oldstate

**Comment by MichelML** [comment]
@sroy3 https://drive.google.com/file/d/1kbEAJRdsOyOJhUsbKcOSI1c7ASUN_AmT/view ... a slideIn, slideDown

i started from here https://github.com/frankwallis/react-slidedown/tree/master/lib and reimplemented it with the new API of the lib you suggested

File: src/components/collapsibleContainer/CollapsibleContainerReducers.spec.ts (hunk 7 above)
Comment (MichelML) [reply src/components/collapsibleContainer/CollapsibleContainerReducers.spec.ts]: somewhat done depending on what you meant...you can check the last test

File: src/components/collapsibleContainer/CollapsibleContainer.tsx [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    return (
+      <div id={this.props.id} className={classNames('collapsible', this.props.className)}>
+        <button
+          type='button'
+          className={this.getCollapsibleHeaderClass()}
+          onClick={(e: React.MouseEvent<HTMLInputElement>) => this.props.onToggleExpandedState(this.props.expanded)}>
+          {this.props.title}
+          {this.getCollapsibleHeaderIcon()}
+        </button>
+        <SlideY in={this.props.expanded} timeout={350}>
```
Comment (GermainBergeron) [inline src/components/collapsibleContainer/CollapsibleContainer.tsx]: Really nice, I like it!

**Pull request merged**